File: pset6/credit.py
##
# Validates a credit card number
# according to algorithm invented by Hans Peter Luhn
##

import cs50


def main():
    print("Number: ", end="")
    card = cs50.get_int()
    sum = sum_flip(card, False, 2) + sum_flip(card, True, 1)
    if (sum % 10) == 0:
        i = card
        while i > 10:
            i //= 10
        m = get_sec(card)
        if i == 3:
            if m == 4 or m == 7:
                print("AMEX")
            else:
                print("INVALID")
        elif i == 4:
            print("VISA")
        elif i == 5:
            if m > 0 and m < 6:
                print("MASTERCARD")
            else:
                print("INVALID")
        else:
            print("INVALID")
    else:
        print("INVALID")
    exit(0)


def get_sec(card):
    while True:
        if card <= 100:
            break
        card //= 10
    return int(card % 10)


def sum_flip(i, flip, m):
    sum = 0
    while i != 0:
        if flip:
            ss = sum_sum(int(i % 10)*m)
            sum += ss
        i //= 10
        flip = not flip
    return sum


def sum_sum(i):
    sum = 0
    while i != 0:
        sum += i % 10
        i //= 10
    return sum

if __name__ == "__main__":
    main()

File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>

int get_sec(long long i);
int sum_flip(long long i, bool flip, int m);
int sum_sum(long long  i);


int main(void)
{
    printf("Number: ");
    long long card = get_long_long();
    int sum = sum_flip(card,false,2) + sum_flip(card,true,1);
    if (sum % 10 == 0)
    {
        long long i = card;
        while (i>10)
        {
            i /=10;
        }
        int m = get_sec(card);
        switch (i)
        {
            case 3:
                if (m == 4 || m == 7)
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");
                }
                break;
            case 4:
                printf("VISA\n");
                break;
            case 5:
                if (m > 0 && m < 6)
                {
                    printf("MASTERCARD\n");
                }
                else
                {
                    printf("INVALID\n");
                }
                break;
            default:
                printf("INVALID\n");
                break;
        }
    }
    else 
    {
        printf("INVALID\n");
    }
}

int get_sec(long long i)
{
    while (i>100)
    {
        i /= 10;
    }
    return (int) (i%10);
}

int sum_flip(long long i, bool flip, int m)
{
    int sum = 0;
    while (i != 0)
    {
        if (flip)
        {
            sum+= sum_sum((i%10)*m);
        }
        i /= 10;
        flip = !flip;
    }
    return sum;
}

int sum_sum(long long  i)
{
    int sum = 0;
    while (i != 0)
    {
        sum+= i%10;
        i /= 10;
    }
    return sum;
}